### DISCUSSION TITLE
Azure Function - minute+ delays on call to Language.from_bytes command while running multiple worker processes

### DISCUSSION THREAD

**Discussion by bwholman** [explosion/spaCy#6499] [Help: Coding & Implementations] [answered]
* Operating System: function deployed to linux based docker profile provided by MS  -- Local dev Windows 10 - no issues there
* Python Version Used: 3.8.6
* spaCy Version Used: nightly
* Environment Information:  the model I'm loading is a custom one I created and saved using nlp.to_bytes() and is 7,056 kb.  working well in the windows dev machine situation.

The code below works perfectly when running this azure function locally on Windows 10 using "func start" with FUNCTIONS_WORKER_PROCESS_COUNT=10.  I can throw many requests at the local REST api and they are all run spacy nlp processes concurrently.  This is the performance we are looking for in our deployed azure function version.

This function also works when a single function call is made to the azure deployed code, albeit the from_bytes model load takes approx 30 seconds.

The issue arises when it is deployed to an azure function and called to run multiple requests concurrently, the following code comes to a grinding halt on the line "nlp.from_bytes": (this is in an async class method...)
>self.file_contents = bytes()
>async for chunk in self.__bytes_from_file(model_location):
>>self.file_contents += bytearray(chunk)

>logging.info("model load from bytes started - this is where the delay occurs")
>self.nlp = nlp.from_bytes(self.file_contents)
>logging.info("model loaded from bytes")
>self.__build_span_extensions()
>.
>.
>.
>async def __bytes_from_file(self, filename, chunksize=16384):
>>async with aiofiles.open(filename, "rb") as f:
>>>while True:
>>>>chunk = await f.read(chunksize)
>>>>if chunk:
>>>>>yield chunk
>>>>else:
>>>>>break

I'm investigating this from the azure hosting side as well, but at this point I'm floating it by the spacy experts to see if you've encountered something similar ... Plus I've never been accused of being a python expert so maybe I'm doing something dumb?  

Thanks all.

**Comment by bwholman** [comment]
Here is a sample log file from our azure function instance - just one run, no CPU memory competition (10070 is the process id).  Note the .from_bytes process 1min26sec, and that number goes up with concurrent processes happening.
2020-12-04T00:25:57Z   [Information]   10070:  model read starting
2020-12-04T00:25:58Z   [Information]   10070:  model read complete: 7225128 bytes
2020-12-04T00:**25:58**Z   [Information]   10070:  model from bytes starting
2020-12-04T00:**27:24**Z   [Information]   10070:  model from bytes complete: ['entity_ruler', 'sentencizer']

Here is a sample log file from the exact same code running on my laptop Win10.  7 seconds, and that stays at 7 seconds when multiple processes concurrently run.
[2020-12-04T00:19:55.089Z] 6584:  model read starting
[2020-12-04T00:19:55.860Z] 6584:  model read complete: 7225128 bytes
[2020-12-04T00:19:**55**.906Z] 6584:  model from bytes starting
[2020-12-04T00:20:**02**.614Z] 6584:  model from bytes complete: ['entity_ruler', 'sentencizer']

**Comment by adrianeboyd** [comment]
I'm not sure I know enough about the details, but a couple things:

* Are you sure the async file io is beneficial here? Where are you reading the model from in azure? What happens if you just use `spacy.load()` instead? The model is small enough that I doubt that there's any benefit from loading it this way. And especially, why force it to read the file in such small chunks?

* I'm not sure I understand all the details of your setup, but if you are loading the model at the start of each request, it would probably be better to restructure things so that the model is loaded once on initialization and stays in memory so you're not waiting 10-30 s for this for each request. But whether this makes sense can depend on your setup, of course.

**Comment by bwholman** [comment, marked as answer]
right you are... I was trying to fit this into an Azure Function which was NOT the correct approach.  just put together a small flask driven api, loaded the model on app startup and it's working great/fast as an azure App Service.  (no async, using `spacy.load()`)
